Add OBJ model support

Talon1024 opened the pull request:
This will allow GZDoom to use OBJ models in addition to MD2, MD3, DMD, and UE1 models.

Pros:
- No limits imposed by the file format
- Text format; Easily editable in a text editor
- Supported by most 3D programs and libraries

Cons:
- Static (no animation support)
- Files can be very large

As of the latest commit, I'm mostly satisfied. The code has been tested in all of the renderers, including the software renderers. I've squashed my commits so it is more "useful". However, there are some things/features that are allowed in OBJ which I don't consider very important, such as faces with more than 4 sides.

raa-eruanna commented:
First things first - do you have any test material for this? What you used to code it would be preferable - but also any other source of OBJ files as well - a diversity of different exports so we know what ones work and what ones don't (if any, hopefully not any).

Secondly - @alexey-lysiuk tends to like pull requests to be squashed. I am not sure if that's going to be a requirement here, but I predict it will help.

Talon1024 commented:
[This zip](http://www.ciinet.org/kevin/myimages2/doom/misc/OBJModels.zip) was the mod I used to test/debug this code and get it working as I wanted it to.

I already squashed a few of my commits. Before I added [d9405c8](https://github.com/coelckers/gzdoom/pull/502/commits/d9405c873a8db8e71834432f53941857a8e8e75f), there were 6 commits, and I squashed some of the smaller ones into the bigger ones.

raa-eruanna commented:
I got a crash when attempting to use this model here - https://www.turbosquid.com/FullPreview/Index.cfm/ID/1093267

This is the top of the stack trace:
`>	gzdoom.exe!FOBJModel::BuildVertexBuffer(FModelRenderer * renderer) Line 282	C++`

Talon1024 commented on src/r_data/models/models_obj.cpp:
```diff
@@ -187,32 +187,47 @@ void FOBJModel::ParseFaceSide(const FString &sideStr, OBJFace &face, int sidx)
 	if (sideStr.IndexOf("/") >= 0)
 	{
-		TArray<FString> sides = sideStr.Split("/");
+		TArray<FString> sides = sideStr.Split("/", FString::TOK_KEEPEMPTY);
```
I'll have to change each forward slash to something else because having two forward slashes together is valid OBJ syntax, but FScanner recognizes that as a line comment.

Talon1024 commented:
OK, I think this is good to go. The previous model you got from TurboSquid crashes the software renderers, though, but I tested my code with other high-poly models, such as an export of [The Given RC2](https://www.doomworld.com/forum/topic/95543-the-given-a-large-monsterless-puzzle-map-on-idgames/?tab=comments#comment-1783913) MAP01 from GZDoom Builder, and a model of a mountain range from WolfenDoom: Blade of Agony C3M5_C exported by GZDoom Builder as well.

Talon1024 commented:
Done. The compilation errors had nothing to do with my code, however.

coelckers commented:
What's the status here?

Talon1024 commented:
- The [OBJ model from TurboSquid](https://github.com/coelckers/gzdoom/pull/502#issuecomment-394461739) works.
- Exported models of levels (created with GZDoom Builder) from Freedoom, Ancient Aliens, and The Given RC2 all work.
- OBJ model rendering works in both software renderers.
- OBJ models that have faces with more than 4 sides have holes in them, although I see this feature of OBJ as unimportant, as it would be a pain in the ass to implement the necessary code (ear clipping) to properly triangulate those kinds of faces, and most OBJs I've seen either use triangular or quad faces.
- Faces without UV and/or normal references are properly handled.
- Faces with negative UV, normal, and/or vertex references are properly handled.

alexey-lysiuk commented:
Looks good. Merging this PR will allow to get some real feedback from modders.
